scripts/remoteperformance.ps1

```diff
@@ -0,0 +1,567 @@
+}
+
+# Join path in script to ensure right platform separator
+$RemoteDirectory = Invoke-Test-Command -Session $session -ScriptBlock { Join-Path (Get-Location) "Tests" }
+
+$LocalDirectory = Join-Path $RootDir "artifacts"
+
+if ($Local) {
+    $RemoteDirectory = $LocalDirectory
+}
+
+function WaitFor-Remote-Ready {
+    param ($Job, $Matcher)
+    $StopWatch =  [system.diagnostics.stopwatch]::StartNew()
+    while ($true) {
+        $CurrentResults = Receive-Job -Job $Job -Keep
+        if (![string]::IsNullOrWhiteSpace($CurrentResults)) {
+            $DidMatch = $CurrentResults -match $Matcher
+            if ($DidMatch) {
+                return $true
+            }
+        }
+        Start-Sleep -Seconds 0.1 | Out-Null
+        if ($StopWatch.ElapsedMilliseconds -gt 10000) {
+            return $false
+        }
+    }
+}
+
+function WaitFor-Remote {
+    param ($Job)
+    Wait-Job -Job $Job -Timeout 10 | Out-Null
+    $RetVal = Receive-Job -Job $Job
+    return $RetVal -join "`n"
+}
+
+function Copy-Artifacts {
+    param([string]$From, [string]$To)
+    try {
+        Invoke-Test-Command $session -ScriptBlock { param($To) Remove-Item -Path "$To/*" -Recurse -Force } -ArgumentList $To
+    } catch {
+        # Ignore failure
+    }
+    # TODO Figure out how to filter this
+    Copy-Item -Path "$From\*" -Destination $To -ToSession $session  -Recurse
+}
+
+class ArgumentsSpec {
+    [string]$All;
+    [string]$Loopback;
+    [string]$Remote;
+
+    [string]GetArguments() {
+        if ($script:Local) {
+            return "$($this.All) $($this.Loopback)"
+        } else {
+            return "$($this.All) $($this.Remote)"
+        }
+    }
+}
+
+class ExecutableSpec {
+    [string]$Platform;
+    [string[]]$Tls;
+    [string[]]$Arch;
+    [string]$Exe;
+    [ArgumentsSpec]$Arguments;
+}
+
+class TestDefinition {
+    [string]$TestName;
+    [ExecutableSpec]$Remote;
+    [ExecutableSpec]$Local;
+    [int]$Iterations;
+    [string]$RemoteReadyMatcher;
+    [string]$ResultsMatcher;
+
+    [string]ToString() {
+        $RetString = ("{0}_{1}_{2}_{3}" -f $this.TestName,
+                                        $this.Remote.Platform,
+                                        $script:RemoteTls,
+                                        $script:RemoteArch
+                                        )
+        if ($script:Local) {
+            $RetString += "_Loopback"
+        }
+        return $RetString
+    }
+
+    [string]ToTestPlatformString() {
+        $RetString = ("{0}_{1}_{2}" -f    $this.Remote.Platform,
+                                    $script:RemoteTls,
+                                    $script:RemoteArch
+                                         )
+        if ($script:Local) {
+            $RetString += "_Loopback"
+        }
+        return $RetString
+    }
+}
+
+class TestPublishResult {
+    [string]$PlatformName
+    [string]$TestName
+    [string]$CommitHash
+    [double[]]$IndividualRunResults
+}
+
+$currentLoc = Get-Location
```

Comment:
This script scatters a lot of stuff at the global execution scope throughout. Can we clean it up some? If there are global variables that need to be initialized for everything, put them at the top. Otherwise all the global execution goes as the end?